### PR TITLE
HIVE-25064

### PR DESCRIPTION
**Pull request by szlta**
- create separate shader maven module for Iceberg libraries
- fix transitive dependency issue (e.g. caffeine version 2.8.4 is bundled in the shaded jar instead of 2.7.0)
- allow easier debugging of iceberg-handler and iceberg-catalog modules
- allow referencing two separate versions of libraries that both Hive and Iceberg use, such as Parquet,Avro,Orc..

### CODE REVIEW

**Review by lcspinter — APPROVED** [review]

File: iceberg/iceberg-catalog/pom.xml [hunk 1]
```diff
@@ -21,9 +21,9 @@
 
     <dependencies>
         <dependency>
-            <groupId>org.apache.iceberg</groupId>
-            <artifactId>iceberg-core</artifactId>
-            <optional>true</optional>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-iceberg-shading</artifactId>
+            <version>${project.parent.version}</version>
```
Comment (pvary) [inline iceberg/iceberg-catalog/pom.xml:26]: Could we move this to the parent pom?

File: iceberg/iceberg-handler/pom.xml [hunk 2]
```diff
@@ -120,95 +93,6 @@
     </dependencies>
 
     <profiles>
```
Comment (pvary) [inline iceberg/iceberg-handler/pom.xml:95]: Do we need the `<profiles>` here?

File: iceberg/iceberg-shading/pom.xml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        <hive.path.to.root>../..</hive.path.to.root>
+        <path.to.iceberg.root>..</path.to.iceberg.root>
+        <shade.relocation.package.prefix>org.apache.hive.iceberg</shade.relocation.package.prefix>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.iceberg</groupId>
+            <artifactId>patched-iceberg-core</artifactId>
+            <version>patched-${iceberg-api.version}-${project.parent.version}</version>
```
Comment (pvary) [inline iceberg/iceberg-shading/pom.xml:45]: Could we move this to the parent pom?

**Comment by pvary** [comment]
If this solves the compilation problem in the IntelliJ as well, then we can remove the `iceberg` flag introduced by #2188.
Or shall we revert it in a different PR?

**Review by pvary — APPROVED** [review]

File: iceberg/pom.xml [hunk 4]
```diff
@@ -51,14 +52,24 @@
...
             <dependency>
                 <groupId>org.apache.iceberg</groupId>
-                <artifactId>iceberg-core</artifactId>
+                <artifactId>patched-iceberg-core</artifactId>
                 <version>patched-${iceberg-api.version}-${project.parent.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.apache.iceberg</groupId>
+                <artifactId>iceberg-api</artifactId>
+                <version>${iceberg-api.version}</version>
```
Comment (marton-bod) [inline iceberg/pom.xml:66]: not completely related to this PR, but why isn't this just `iceberg.version`? I guess there shouldn't be different version for iceberg-api, iceberg-data, etc.

File: iceberg/iceberg-shading/pom.xml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                                        </relocation>
+                                        <relocation>
+                                            <pattern>com.github.benmanes</pattern>
+                                            <shadedPattern>${shade.relocation.package.prefix}.com.github.benmanes</shadedPattern>
+                                        </relocation>
+                                    </relocations>
+                                    <artifactSet>
+                                        <includes>
+                                            <include>org.apache.iceberg:*</include>
+                                            <include>org.apache.hive:hive-iceberg-catalog</include>
```
Comment (marton-bod) [inline iceberg/iceberg-shading/pom.xml:118]: if the catalog module depends on the shading module, isn't it a problem that the shading module's shade step depends on the catalog? Is this some form of a circular dependency? Maybe not, just checking :)

**Comment by marton-bod** [comment]
@szlta Have you been able to double-check if the correct jars are packaged into the `packaging/target/..../lib` folder after this change? In theory, it should only be the fat handler jar

**Comment by marton-bod** [comment]
Actually, now that I'm thinking about it, we are currently also including the iceberg-catalog jar into the packaging/lib, which would mean we have the catalog related classes twice on the classpath (and probably in an undeterminable order).

File: iceberg/iceberg-shading/pom.xml (hunk 5 above)
Comment (szlta) [reply iceberg/iceberg-shading/pom.xml:118]: Ah good catch that's a benign leftover, it's a noop whitelist entry in this case, as the related classes are not there yet during the build time of iceberg-shading. I'll remove it

File: iceberg/pom.xml (hunk 4 above)
Comment (szlta) [reply iceberg/pom.xml:66]: Hm yeah it should be iceberg.version everywhere, but I was just following suit with this one, as I see the -api version being used in all poms. Maybe we could address this in separate jira.

Comment (szlta) [reply iceberg/pom.xml:66]: On 2nd thought I'll just go ahead and fix it in the next iteration of this change. I'll keep iceberg-api.version but that will only be applicable to -api lib.

**Review by marton-bod — APPROVED** [review]

**Pull request merged**